### ISSUE TITLE
Feature Request: Add ability to specify file extension for parsers

### ISSUE THREAD

**Issue by jasonlyle88** [liquibase/liquibase#4389]
### Search first

- [X] I searched and no similar issues were found

### Description

Right now Liquibase ships with a defined set of parsers: XML, YAML, JSON, and SQL.

As of right now, each of those parsers have a hard coded and pre-defined set of file extensions that are recognized for that parser:

- XML: `.xml`
- YAML: `.yaml` (and I believe `.yml` as well)
- JSON `.json` (and maybe `.js` as well?)
- SQL: `.sql`

It would be nice to be able to control this list, adding or removing extensions as desired for the project.

Why I see this as helpful is for projects (like mine) that are largely managed with XML changelogs for ordering, and taking advantage of includeAll attributes to include entire directories of SQL changelogs.

In this type of project structure, there are two major advantages: removing extensions from parsers we don't want to use for our project and adding more extensions for the SQL parser.

Going into detail, I'll explain the first point of removing extensions. In my example, we don't want anything that is not a SQL changelog or an XML changelog to be considered in the search. So, by removing the extensions from the yaml and json parser, we could include those file. For example, when including a front end javascript file (or having database data that is represented in a JSON format and stored in JSON files), includeAll will pick up these files and not know how to parse them. So being able to turn those file extensions off when not intended to be used in a project would be very nice!

As for the second item, I'd like to present the following example. In the case of projects utilizing mostly just includeAll statements for SQL changelogs, we usually have to spit the execution of package specs and package bodies. For example, we have to compile package specs, then views (because a view may want to call a method in a package spec, but a package body may need to utilize the view). In this case, package bodies and package specs are usually put in the same directory (since its easiest to edit them and switch back and forth between them this way). This is accomplished by the file being names package_name.pks for a package spec and package_name.pkb for a package body. Or having files with the extension .pls for plsql files and .sql for just sql files. And any other number of extensions that I have seen over the years. Being able to say that these file extensions should be parsed by the following parser would greatly help in these type of situations!

If this is considered, I would then like to submit a separate feature request for a new attribute on includeAll that allows filtering by file extension to allow for more useful project structuring (without having to write an entire java class to filter).

Happy to discuss this idea and implementation ideas! I can potentially also build a PR for this and for the includeAll file extension filter attribute.

### Steps To Reproduce

Feature request to add functionality, no reporduction possible

### Expected/Desired Behavior

See above description

### Liquibase Version

All

### Database Vendor & Version

All

### Liquibase Integration

CLI

### Liquibase Extensions

None

### OS and/or Infrastructure Type/Provider

All

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

**Comment by MalloD12** [comment]
Hi @jasonlyle88,

This sounds like a good idea, but to be honest I'm not sure to be fully understanding one of the reasons you mentioned on this request.  I think it is a good idea to add support for allowing users to include some custom parser/s which might be useful for their purposes, and I think maybe we can do this by adding a new config for `ChangeLogParserConfiguration`:

```
public static final ConfigurationDefinition<String[]> USER_CHOSEN_PARSERS;

USER_CHOSEN_PARSERS = builder.define("userChosenParsers", String[].class)
                .setDescription("User chosen parsers to allow changelog be processed for the given format type/s")
                .build();
```

And then register the given parsers with `ChangeLogParserFactory.register(List<ChangeLogParser> changeLogParsers)` (and maybe unregister any other out of that list).

What I don't see yet is what would be the benefit of removing some of the pre-existent parsers if someone doesn't need them in their project. How they could impact negatively if they co-exist with the ones that are useful for a given project?

Thanks,
Daniel.

**Comment by jasonlyle88** [comment]
Hey @MalloD12 ,

We don't necessarily need to remove the parser, just "turn it off" or make sure it doesn't parse certain extensions. In our project we are only using changelogs in the xml and sql format... but we do have json files that contain json for other purposes in our project. As of right now, these files have to be put in a different part of the project which is not covered by an includeAll because the includeAll will pickup the json file (even though it isn't a changelog) and attempt to parse it and fail. I feel like I'm just restating what I said above... so I'm not sure if this clears anything up or not.  If it doesn't, please let me know and I will setup a small directory example and put the tree and files contents here to help demonstrate the circumstance I'm describing

**Comment by MalloD12** [comment]
I think I'm seeing better what you mean. Considering your need, wouldn't this purpose be achieved once having merged your `endsWithFilter` [PR](https://github.com/liquibase/liquibase/pull/4402)? Plus, having a sort of configuration like the one I mentioned previously to unregister the parsers you guys don't need?

**Comment by jasonlyle88** [comment]
Yes, 100% it will be achievable with my ensWithFilter PR (would just have to be applied on all necessary includeAll statements) or with the configuration (only have to setup once 🎉 ) you mentioned! I really like your idea, it is along the lines of what I was thinking. I even discovered the ChangeLogParserFactory.register... it was just the ConfigurationDefinition part I couldn't seem to figure out
